core/templates/services/rte-helper-modal.controller.ts

```diff
@@ -311,6 +324,23 @@ export class RteHelperModalComponent {
     return url !== text;
   }
 
+  disableSaveButtonForVideoRte(): void {
```

Comment:
Can you generalize this to a disableSaveButtonForRte() and have the interior logic use this.componentId to determine which validation rule apply and handle the validation accordingly?

Also I feel like this is misnamed because it isn't actually doing the disabling -- I would expect calling this function to result in the button being disabled. But I think you're just connecting a subscription here. Suggest renaming to be clearer about what the function is doing.

Reply:
@seanlip As I have exams coming up I do not have the time to generalize this into a complete disableSaveButtonForRte() so renaming this into a more appropriate name and using the existing logics might be sufficient to handle this issue?

Reply:
It is not hard to generalize this function, and given that validation is a standard use case I think it is worth doing. In general the logic in this controller needs to be RTE-component-id-agnostic, which means that you should not have functions with "videoRte", "linkRte", etc. in them. As you can see from the other discussions, we're trying to get rid of that sort of thing, so it's a bit counterproductive to add more of that with this PR :)

If you need to do this after exams, it is probably fine, just let us know when you'll be away from and to so that we don't assign the issue to someone else. Thanks!

Reply:
@seanlip I see, since the validation is based on the input of the fields I think it is reasonable placing a subscription on the relevant fields for all the cases. This is the redone disableSaveButtonForRte() that uses the componentId to decide which validation to run
```ts
  disableSaveButtonForRte(): void {
    if (this.componentId === this.COMPONENT_ID_MATH) {
      this.formSubscription = this.customizationArgsForm.valueChanges.subscribe(
        (value) => {
          let rawLatex: string = value[0].raw_latex;
          let mathExpressionSvgIsBeingProcessed:
          boolean = value[0].mathExpressionSvgIsBeingProcessed;
          this.saveButtonIsDisabled =
          mathExpressionSvgIsBeingProcessed ||
          rawLatex === '';
        });
    } else if (this.componentId === this.COMPONENT_ID_VIDEO) {
      this.formSubscription = this.customizationArgsForm.valueChanges.subscribe(
        (value) => {
          let start: number = value[1];
          let end: number = value[2];
          if (start === 0 && end === 0) {
            this.saveButtonIsDisabled = false;
          } else {
            this.saveButtonIsDisabled = start >= end;
          }
        }
      );
    } else if (this.componentId === this.COMPONENT_ID_LINK) {
      this.formSubscription = this.customizationArgsForm.valueChanges.subscribe(
        (value) => {
          let url: string = value[0];
          let text: string = value[1];
          // First check if the `text` looks like a URL.
          const suffixes = ['.com', '.org', '.edu', '.gov'];
          for (const suffix of suffixes) {
            if (url.endsWith(suffix)) {
              this.urlSuffixIsInvalid = false;
            }
          }
          if (!this.urlSuffixIsInvalid) {
            // The link text. If left blank, the link URL will be used.
            if (text === '') {
              text = url;
              this.saveButtonIsDisabled = false;
            }
            // If the text looks like a URL, strip the leading 'http://' or
            // 'https://' or 'www.'.
            const prefixes = ['https://', 'http://', 'www.'];
            for (const prefix of prefixes) {
              if (url.startsWith(prefix)) {
                url = url.substring(prefix.length);
              }
              if (text.startsWith(prefix)) {
                text = text.substring(prefix.length);
              }
            }
            // After the cleanup, if the strings are not equal, then we do not
            // allow the lesson creator to save it.
            this.saveButtonIsDisabled = url !== text;
          }
        }
      );
    }
  }
```
which is then used in combination with the html as follows
```html
<div class="oppia-customization-arg-form-container">
  <div class="modal-header">
    <h3>Customize This Component</h3>
  </div>

  <div class="modal-body">
    <!-- Provide a sink to move the focus into while the modal loads. -->
    <input *ngIf="modalIsLoading" type="text" [oppiaFocusOn]="tmpFocusPoint">
    <form *ngIf="customizationArgsForm" [formGroup]="customizationArgsForm">
      <div *ngFor="let customizationArgSpec of customizationArgSpecs;let ind = index"
           class="oppia-customization-arg-editor">
        <strong>{{ customizationArgSpec.description }}</strong>
        <div class="oppia-customization-arg-editor-inner">
          <schema-based-editor [formControlName]="ind"
                               [schema]="customizationArgSpec.schema">
          </schema-based-editor>
        </div>
      </div>
    </form>
    <div *ngIf="urlSuffixIsInvalid && componentId === 'link'" class="link-error-message">
      Please enter a URL ending with '.com', '.org', '.edu', '.gov'. 
    </div>
    <div *ngIf="saveButtonIsDisabled && componentId === 'link'" class="link-error-message">
      It seems like clicking on this link will lead the user to a different URL
      than the text specifies. Please change the text.
    </div>
    <div *ngIf="saveButtonIsDisabled && componentId === 'video'" class="link-error-message">
      Please ensure that the start time of the video is earlier than the end time.
    </div>
  </div>

  <div class="modal-footer">
    <input type="submit"
           class="btn btn-danger e2e-test-delete-button"
           value="Delete"
           *ngIf="!defaultRTEComponent"
           (click)="delete()">
    <input class="btn btn-secondary e2e-test-cancel-rich-text-editor oppia-cancel-btn"
           value="Cancel"
           (click)="cancel()">
    <input *ngIf="componentId !== 'link' && componentId !== 'math' && componentId !== 'video'"
           class="btn btn-success e2e-test-close-rich-text-component-editor oppia-save-btn"
           value="Done"
           (click)="save()"
           [disabled]="!customizationArgsForm.valid">
    <input *ngIf="componentId === 'link'"
           class="btn btn-success e2e-test-close-rich-text-component-editor oppia-save-btn"
           value="Done"
           (click)="save()"
           [disabled]="!customizationArgsForm.valid || disableSaveButtonForRte() || saveButtonIsDisabled || urlSuffixIsInvalid">
    <input *ngIf="componentId === 'math'"
           class="btn btn-success e2e-test-close-rich-text-component-editor oppia-save-btn"
           value="Done"
           (click)="save()"
           [disabled]="!customizationArgsForm.valid || disableSaveButtonForRte() || saveButtonIsDisabled">
    <input *ngIf="componentId === 'video'"
           class="btn btn-success e2e-test-close-rich-text-component-editor oppia-save-btn"
           value="Done"
           (click)="save()"
           [disabled]="!customizationArgsForm.valid || disableSaveButtonForRte() || saveButtonIsDisabled">
  </div>
</div>
<style>
  .oppia-customization-arg-form-container .oppia-customization-arg-editor {
    margin-bottom: 25px;
    z-index: 3001;
  }
  .oppia-customization-arg-form-container
    .oppia-customization-arg-editor-inner {
    margin-top: 15px;
  }
  .oppia-customization-arg-form-container .link-error-message {
    color: #f00;
    display: block;
    font-size: 0.8em;
  }
  .oppia-customization-arg-editor-modal {
    z-index: 1;
  }
  input.oppia-save-btn, input.oppia-cancel-btn {
    max-width: 100px
  }
</style>

```
to get the desired validation behaviour. I think this also address another comment regarding not using different this.currentRteIsXEditor and only the componentId that is already set as it is passed to the controller. Please let me know what you think